src/Components/Blazor/Build/src/targets/Blazor.MonoRuntime.targets

```diff
@@ -219,49 +219,110 @@
 
     <ItemGroup>
-      <BlazorOutputWithTargetPath Include="@(_BlazorResolvedRuntimeDependencies)">
-        <TargetOutputPath>$(BlazorRuntimeBinOutputPath)%(FileName)%(Extension)</TargetOutputPath>
-      </BlazorOutputWithTargetPath>
+      <_BlazorProjectReference Include="@(ReferencePath->WithMetadataValue('ReferenceSourceTarget', 'ProjectReference'))" />
+      <_BlazorExternalReference Include="@(ReferencePath)" Exclude="@(_BlazorProjectReference)" />
     </ItemGroup>
+
+    <BlazorGetAssemblyReferences Assemblies="@(IntermediateAssembly);@(_BlazorProjectReference)">
+      <Output TaskParameter="AssemblyReferences" ItemName="_ProjectAssemblyReferences" />
+    </BlazorGetAssemblyReferences>
+
+    <Hash ItemsToHash="@(_BlazorProjectReference);@(_ProjectAssemblyReferences);">
+      <Output TaskParameter="HashResult" PropertyName="_BlazorReferenceHash" />
+    </Hash>
+
+    <WriteLinesToFile
+      Lines="$(_BlazorReferenceHash)"
+      File="$(_BlazorReferenceCacheFile)"
+      Overwrite="True"
+      WriteOnlyWhenDifferent="True" />
   </Target>
 
   <Target
-    Name="_ResolveBlazorRuntimeDependencies"
+    Name="_GetBlazorNotLinkedAssemblyClosure"
+    DependsOnTargets="_GenerateBlazorLinkerDescriptor"
     Inputs="$(ProjectAssetsFile);
-            @(IntermediateAssembly);
-            @(_BlazorDependencyInput)"
-    Outputs="$(_BlazorApplicationAssembliesCacheFile)">
+            $(_BlazorReferenceCacheFile);
+            @(_BlazorExternalReference);"
+    Outputs="$(_BlazorLinkerOutputCache)">
+
+   <ItemGroup>
+      <_OldLinkedFile Include="$(BlazorIntermediateLinkerOutputPath)*.dll" />
+      <_OldLinkedFile Include="$(BlazorIntermediateLinkerOutputPath)*.pdb" />
+    </ItemGroup>
+
+    <Delete Files="@(_OldLinkedFile)" />
 
     <!--
-    At this point we have decided not to run the linker and instead to just copy the assemblies
-    from the BCL referenced by the app the nuget package into the _framework/_bin folder.
-    The only thing we need to do here is collect the list of items that will go into _framework/_bin.
+      When running from Desktop MSBuild, DOTNET_HOST_PATH is not set.
+      In this case, explicitly specify the path to the dotnet host.
     -->
-    <ResolveBlazorRuntimeDependencies
-      EntryPoint="@(IntermediateAssembly)"
-      ApplicationDependencies="@(_BlazorDependencyInput)"
-      WebAssemblyBCLAssemblies="@(_WebAssemblyBCLAssembly)">
+    <PropertyGroup Condition=" '$(DOTNET_HOST_PATH)' == '' ">
+      <_DotNetHostDirectory>$(NetCoreRoot)</_DotNetHostDirectory>
+      <_DotNetHostFileName>dotnet</_DotNetHostFileName>
+      <_DotNetHostFileName Condition=" '$(OS)' == 'Windows_NT' ">dotnet.exe</_DotNetHostFileName>
+    </PropertyGroup>
+
+    <BlazorILLink
+        ILLinkPath="$(MonoLinkerPath)"
+        AssemblyPaths="@(_BlazorDependencyInput);@(_WebAssemblyBCLAssembly)"
+        RootAssemblyNames="@(IntermediateAssembly)"
+        RootDescriptorFiles="@(BlazorLinkerDescriptor)"
```

Comment:
Why are we passing a linker descriptor file? If this is only for reference resolution, does the descriptor file contain any useful information?

Reply:
👍 we could probably remove this